### PR TITLE
undo icontentish interface

### LINKED ISSUE #3833: Un-remove IContentish from site root class
On https://github.com/plone/Products.CMFPlone/blob/016459fd9d023017e9dc0a0b635bd66099826db1/Products/CMFPlone/Portal.py#L214 I remove IContentish, because having it broke tests and setup code in places that I didn't feel like touching.

Nowadays we have some test failures because of the removal. See https://github.com/plone/plone.restapi/pull/1674#issuecomment-1664193927 for additional details, but suffice to say that `conversation_view` is expected to exist on the site root, but isn't because it's bound to `IContentish`.

### PR DESCRIPTION
**Pull request by Akshat2Jain**
fixes #3833 

also related to https://github.com/plone/plone.restapi/pull/1674

### CODE REVIEW

File: Products/CMFPlone/Portal.py [hunk 1]
```diff
@@ -222,8 +222,7 @@ def isEffective(self, date):
         return 1
 
 
-# Remove the IContentish interface so we don't listen to events that won't
-# apply to the site root, ie handleUidAnnotationEvent
-classImplementsOnly(PloneSite, implementedBy(PloneSite) - IContentish)
+# Add the IContentish interface back to the PloneSite class
+classImplementsOnly(PloneSite, implementedBy(PloneSite) + IContentish)
```
Comment (jaroel) [inline Products/CMFPlone/Portal.py:226]: You can drop the whole line. `IContentish` was already part of `PloneSite` - I just removed it with `classImplementsOnly`.

If we succesfully no longer remove `IContentish`, we should start seeing test failures involving `handleUidAnnotationEvent` in the coredev tests.

**Comment by davisagli** [comment]
@jenkins-plone-org please run jobs

**Comment by wesleybl** [comment]
The tests are failing with the error:

```
File "/home/jenkins/.buildout/eggs/cp311/Products.CMFUid-4.0-py3.11.egg/Products/CMFUid/UniqueIdAnnotationTool.py", line 86, in handleUidAnnotationEvent
    uidtool.unregister(ob)
```

in this line of this subscriber:

https://github.com/zopefoundation/Products.CMFUid/blob/8ac5c69df3d8d008561b9ace1cb5fc43d3d3bf1d/src/Products/CMFUid/UniqueIdAnnotationTool.py#L86

The error occurs when creating a Plone Site. It looks like the subscriber wants to use the `portal_uidhandler` tool, but it hasn't been created yet. Then [this line](https://github.com/zopefoundation/Products.CMFUid/blob/8ac5c69df3d8d008561b9ace1cb5fc43d3d3bf1d/src/Products/CMFUid/UniqueIdAnnotationTool.py#L79) could test if it exists, looking like this:

```python
if anno_tool is not None and uidtool is not None:
```

**Comment by Akshat2Jain** [comment]
> The tests are failing with the error:
> 
> ```
> File "/home/jenkins/.buildout/eggs/cp311/Products.CMFUid-4.0-py3.11.egg/Products/CMFUid/UniqueIdAnnotationTool.py", line 86, in handleUidAnnotationEvent
>     uidtool.unregister(ob)
> 'NoneType' object has no attribute 'unregister'
> ```
> 
> in this line of this subscriber:
> 
> https://github.com/zopefoundation/Products.CMFUid/blob/8ac5c69df3d8d008561b9ace1cb5fc43d3d3bf1d/src/Products/CMFUid/UniqueIdAnnotationTool.py#L86
> 
> The error occurs when creating a Plone Site. It looks like the subscriber wants to use the `portal_uidhandler` tool, but it hasn't been created yet. Then [this line](https://github.com/zopefoundation/Products.CMFUid/blob/8ac5c69df3d8d008561b9ace1cb5fc43d3d3bf1d/src/Products/CMFUid/UniqueIdAnnotationTool.py#L78) could test if it exists, looking like this:
> 
> ```python
> if anno_tool is not None and uidtool is not None:
> ```

@wesleybl what do I have to do in this?

**Comment by wesleybl** [comment]
@polyester @icemac

@Akshat2Jain have already signed the Plone contributor agreement. Could you add it to zopefoundation, so I can fix this error?

**Comment by Akshat2Jain** [comment]
any progress?

**Comment by polyester** [comment]
I'm not too sure what my role here is. For historical reasons, we need a (separate) contributor agreement for the Zopefoundation repos; that can be signed here: https://sign.plone.org/d/FTU6XN6qrGtoni

Is the question if @icemac should put that in there for him? Or if @Akshat2Jain does it there himself? I'm not at all familiar with the Zope codebase, testing infrastructure, and customs, although @jaroel is I think so maybe he knows.

**Comment by wesleybl** [comment]
@Akshat2Jain please sign the Zopefoundation contributor agreement. This way you can make a PR for Products.CMFUid

**Comment by Akshat2Jain** [comment]
> @Akshat2Jain please sign the Zopefoundation contributor agreement. This way you can make a PR for Products.CMFUid

Done

**Comment by Akshat2Jain** [comment]
@wesleybl I have been added to the Zope Foundation. Please guide what I have to do next.

**Comment by wesleybl** [comment]
@Akshat2Jain  you need to make a PR on [Products.CMFUid](https://github.com/zopefoundation/Products.CMFUid), with what was explained in: https://github.com/plone/Products.CMFPlone/pull/3836#issuecomment-1705806117

**Comment by Akshat2Jain** [comment]
@wesleybl can you run jobs again?

**Comment by wesleybl** [comment]
@Akshat2Jain It's no use running the jobs again. We need a new release of `Products.CMFUid` with the modifications made in https://github.com/zopefoundation/Products.CMFUid/pull/23

**Comment by wesleybl** [comment]
@jenkins-plone-org please run jobs

**Comment by wesleybl** [comment]
@davisagli @jaroel @Akshat2Jain the errors with `Products.CMFUid` are gone.

We now have two tests failing in `plone.api`. Are they:

* [test_find_interface_dict](https://github.com/plone/plone.api/blob/629e01679e90225f2e3164fe8a6b7e6101436e77/src/plone/api/tests/test_content.py#L1001)
* [test_find_interface_dict__include_not_query](https://github.com/plone/plone.api/blob/629e01679e90225f2e3164fe8a6b7e6101436e77/src/plone/api/tests/test_content.py#L1028)

These tests count the number of objects returned by searching the catalog through the `object_provides` index, using the `IContentish` interface. As Plone Site now also implements this interface, the number of objects has increased by one. Therefore, I think it is only necessary to fix the number of objects in the test asserts.

@Akshat2Jain , can you make a PR on `plone.api` fxing these tests?

**Comment by wesleybl** [comment]
@Akshat2Jain can you please make a PR like this in branch 6.0.x?

https://github.com/plone/Products.CMFPlone/tree/6.0.x

**Comment by Akshat2Jain** [comment]
> @Akshat2Jain can you please make a PR like this in branch 6.0.x?
> 
> https://github.com/plone/Products.CMFPlone/tree/6.0.x

and this pr would be on 6.0x branch? @wesleybl

**Comment by wesleybl** [comment]
> this pr would be on the master branch?

in master of https://github.com/plone/plone.api

> and this pr would be on 6.0x branch?

in branch 6.0.x of https://github.com/plone/Products.CMFPlone

**Comment by wesleybl** [comment]
> @Akshat2Jain can you please make a PR like this in branch 6.0.x?

@Akshat2Jain the "like this" is the PR we are in now #3836, bus in branch 6.0.x.

**Comment by wesleybl** [comment]
@jenkins-plone-org please run jobs

**Comment by wesleybl** [comment]
@jaroel @Akshat2Jain we still have a test failing in plone.restapi:

```
'@id'

  File "/srv/python3.12/lib/python3.12/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/srv/python3.12/lib/python3.12/unittest/case.py", line 634, in run
    self._callTestMethod(testMethod)
  File "/srv/python3.12/lib/python3.12/unittest/case.py", line 589, in _callTestMethod
    if method() is not None:
       ^^^^^^^^
  File "/home/jenkins/.buildout/eggs/cp312/plone.restapi-9.1.0-py3.12.egg/plone/restapi/tests/test_site_serializer.py", line 122, in test_site_root_get_request
    self.assertEqual(response.json()["@id"], self.portal.absolute_url())
                     ~~~~~~~~~~~~~~~^^^^^^^
```

See: https://jenkins.plone.org/job/pull-request-6.1-3.12/26/testReport/junit/plone.restapi.tests.test_site_serializer/TestSiteSerializationFunctional/test_site_root_get_request/

The strange thing is that I ran this test locally on Plone 6.1 coredev and it worked. So it's difficult to know what it is.

**Comment by jaroel** [comment]
> @jaroel @Akshat2Jain we still have a test failing in plone.restapi:
> [...]
> The strange thing is that I ran this test locally on Plone 6.1 coredev and it worked. So it's difficult to know what it is.

Yep. https://github.com/zopefoundation/Products.CMFCore/blob/be1c7590677b978bb9ccf6050b7de84b0d86235d/src/Products/CMFCore/explicitacquisition.py#L29 is triggering the `NotFound`.

Turns out that the aq_chain contains the Plone site twice in `plone.rest`.
If you know what causes _that_, I'd be really happy.

**Comment by wesleybl** [comment]
@jaroel did you manage to make this test fail locally? What did you do?

**Comment by jaroel** [comment]
@jenkins-plone-org please run jobs

**Comment by jaroel** [comment]
@wesleybl Sorry, I had a left over change. My buildout.coredev only checks out Products.CMFPlone branch=undo_icontentish and the test passes.

I fetched origin and then merged origin/master in. That should correctly disable that thingy I do in https://github.com/plone/Products.CMFPlone/blob/621edb0851967f033c40dae89ead14c8f4d4250d/Products/CMFPlone/__init__.py#L12-L21

**Comment by wesleybl** [comment]
@jaroel anyway, this doesn't explain why the test fails in Jenkins and works locally. Are any environment variables set in Jenkins?

@gforcada can you help here?

**Comment by jaroel** [comment]
I'm setting SKIP_PTA in https://github.com/plone/plone.rest/blob/5f9898970f871008e84b254bf73ff13f48218bbc/src/plone/rest/tests/test_explicitacquisition.py#L67-L118, which then leaks into our `test_site_serializer.py`.

I can reproduce with `./bin/test -t test_portal_root -t test_site_root_get_request`. Order is important here

**Comment by wesleybl** [comment]
@jaroel so you need to test the environment variable at the end of these tests and set the SKIP_PTA variable again, to restore the default behavior. Can you take care of it?

**Comment by jaroel** [comment]
@wesleybl please look first: https://github.com/plone/plone.rest/pull/168

**Comment by wesleybl** [comment]
@jaroel anyway, it would be good to create a test similar to `test_site_root_get_request`, but setting `SKIP_PTA `to `false`. In the end, it was a good thing the test failed.

**Comment by jaroel** [comment]
@jenkins-plone-org please run jobs

**Comment by wesleybl** [comment]
@jenkins-plone-org please run jobs

@jaroel I think the tests will pass now. But I don't know if we should merge this before fix https://github.com/plone/plone.rest/issues/169. If someone uses `SKIP_PTA = false`, restapi will not work.

**Comment by jaroel** [comment]
@wesleybl No worries, nobody uses that. It's just there so we don't ship it as enabled by default.

File: Products/CMFPlone/Portal.py [hunk 2]
```diff
@@ -221,9 +221,4 @@ def isEffective(self, date):
         # Override DefaultDublinCoreImpl's test, since we are always viewable.
         return 1
 
-
-# Remove the IContentish interface so we don't listen to events that won't
-# apply to the site root, ie handleUidAnnotationEvent
-classImplementsOnly(PloneSite, implementedBy(PloneSite) - IContentish)
```
Comment (wesleybl) [inline Products/CMFPlone/Portal.py:227]: Please remove imports of `classImplementsOnly`, `implementedBy` and `IContentish`

**Review by wesleybl — CHANGES_REQUESTED** [review]

**Comment by wesleybl** [comment]
@jaroel can you please take care of https://github.com/plone/Products.CMFPlone/pull/3856?

**Comment by jaroel** [comment]
> @jaroel can you please take care of #3856?

amanhã, amanhã (I think I used that correctly)

**Comment by wesleybl** [comment]
@jenkins-plone-org please run jobs

File: news/3833.bugfix [hunk 3]
```diff
@@ -0,0 +1 @@
+Make PloneSite have IContentish again. @Akshat2Jain @jaroel
```
Comment (wesleybl) [inline news/3833.bugfix:1]: I think this is a feature. Not a bugfix.

**Review by wesleybl — CHANGES_REQUESTED** [review]

File: news/3833.bugfix (hunk 3 above)
Comment (jaroel) [reply news/3833.bugfix:1]: I don't have a preference, I just need people to be aware.
@mauritsvanrees any preference so we can move forward?

**Comment by jaroel** [comment]
@davisagli how do you feel about landing this IContentish branch?

**Review by davisagli — APPROVED** [review]
Looks good to me. I also looked through the zopefoundation and plone github organizations for references to IContentish, just to see if there was anything that might be surprising, and I think it should be okay (as also indicated by the state of the tests).

I think I would keep it for 6.1 (master) only though and not make this change for 6.0 at this point.

Special thanks to @Akshat2Jain and @wesleybl for your persistence and patience with this one!

File: news/3833.bugfix (hunk 3 above)
Comment (mauritsvanrees) [reply news/3833.bugfix:1]: Feels more like a regression fix to me, so let's stick with bugfix.

**Comment by wesleybl** [comment]
> I think I would keep it for 6.1 (master) only though and not make this change for 6.0 at this point.

If we don't have this in Plone 6.0, it won't be possible to fix https://github.com/plone/plone.restapi/pull/1674 in an easy way, since we don't have a version of restapi just for Plone 6.1

**Comment by wesleybl** [comment]
I think it would be good to have an upgrade step to reindex the Plone Site, since we have indexes that have interfaces.

**Pull request merged**